### PR TITLE
basic features for remove token filter with unit tests

### PR DESCRIPTION
**Pull request by maoning**
This plugin has almost the same interface as the build-in stop token filter from ElasticSearch with the exception of the support for remove_trailing at the moment, see official ES doc for settings: https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-stop-tokenfilter.html

The exact feature of this remove token plugin is documented by its unit test. It treats all the stop words as if they are not passed into the token stream.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <name>elasticsearch-remove</name>
```
Comment (adrianocrestani) [inline pom.xml]: remove is too vague...remove-stopwords or remove-stop-tokens sounds more accurate

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    public final boolean incrementToken() throws IOException {
+        skippedPositions = false;
+
+        while (input.incrementToken())
+        {
+            if (accept()) {
+
+                if (skippedPositions)
+                {
+                    posIncrAtt.setPositionIncrement(posIncrAtt.getPositionIncrement());
```
Comment (adrianocrestani) [inline src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: this looks like a noop to me

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+
+import java.io.IOException;
+
+public class RemoveTokenFilter extends TokenFilter {
+
+    private final CharArraySet stopWords;
+    private boolean skippedPositions;
+    private int currentStartOffset;
```
Comment (adrianocrestani) [inline src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: this is never used

File: src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    @Override
+    public TokenStream create(TokenStream tokenStream) {
+        return new RemoveTokenFilter(tokenStream, stopWords);
+    }
+
+    public Set<?> stopWords() {
+        return stopWords;
+    }
+
+    public boolean ignoreCase() {
```
Comment (adrianocrestani) [inline src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java]: never used

File: src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+    public RemoveTokenFilterFactory(IndexSettings indexSettings, Environment env, String name, Settings settings) {
+        super(indexSettings, name, settings);
+        this.ignoreCase = settings.getAsBoolean("ignore_case", false);
+        this.stopWords = Analysis.parseStopWords(env, settings, StopAnalyzer.ENGLISH_STOP_WORDS_SET, ignoreCase);
+    }
+
+    @Override
+    public TokenStream create(TokenStream tokenStream) {
+        return new RemoveTokenFilter(tokenStream, stopWords);
```
Comment (adrianocrestani) [inline src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java]: don't you need to pass ignroeCase to the filter?

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java (hunk 3 above)
Comment (maoning) [reply src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: One last thing from offset change I forgot to remove, fixing it.

File: pom.xml (hunk 1 above)
Comment (maoning) [reply pom.xml]: make sense

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java (hunk 2 above)
Comment (maoning) [reply src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: The purpose of skippedPosistions variable is to make sure we only ever increment the token position once no matter how many stop tokens there are consecutively.

File: src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java (hunk 4 above)
Comment (maoning) [reply src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java]: you are right, I'm removing this method. I copied it from the build-in stop analyzer, the accessors were never referenced.

File: src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java (hunk 5 above)
Comment (maoning) [reply src/main/java/elasticsearch.remove/RemoveTokenFilterFactory.java]: Yes :)

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java (hunk 2 above)
Comment (maoning) [reply src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: As discussed offline,  the set(get()) operation is redundant, I'm removing them.

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+import java.io.IOException;
+
+public class RemoveTokenFilter extends TokenFilter {
+
+    private final CharArraySet stopWords;
+    private boolean skippedPositions;
+
+
+    private final PositionIncrementAttribute posIncrAtt = addAttribute(PositionIncrementAttribute.class);
```
Comment (adrianocrestani) [inline src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: no longer used

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+import org.apache.lucene.analysis.TokenStream;
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+
+import java.io.IOException;
+
+public class RemoveTokenFilter extends TokenFilter {
+
+    private final CharArraySet stopWords;
+    private boolean skippedPositions;
```
Comment (adrianocrestani) [inline src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: no longer used

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java (hunk 6 above)
Comment (maoning) [reply src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: fixed :)

File: src/main/java/elasticsearch.remove/RemoveTokenFilter.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import org.apache.lucene.analysis.tokenattributes.CharTermAttribute;
+import org.apache.lucene.analysis.tokenattributes.PositionIncrementAttribute;
+
+import java.io.IOException;
+
+public class RemoveTokenFilter extends TokenFilter {
+
+    private final CharArraySet stopWords;
+
+    private final PositionIncrementAttribute posIncrAtt = addAttribute(PositionIncrementAttribute.class);
```
Comment (adrianocrestani) [inline src/main/java/elasticsearch.remove/RemoveTokenFilter.java]: unused :)

**Pull request merged**